onecodex/stats.py

```diff
@@ -0,0 +1,288 @@
+    ) -> AlphaDiversityStatsResults:
+        """Perform a test for significant differences between groups of alpha diversity values.
+
+        The following tests are supported:
+        - Wilcoxon (2 groups, paired data)
+        - Mann-Whitney U (2 groups, unpaired data)
+        - Kruskal-Wallis with optional posthoc Dunn test (>=2 groups, unpaired data)
+
+        Parameters
+        ----------
+        group_by : str or tuple of str
+            Metadata variable to group samples by. At least two groups are required. If `group_by`
+            is a tuple, field values are joined with an underscore character ("_").
+        paired_by : str or tuple of str, optional
+            Metadata variable to pair samples in each group. May only be used with `test="wilcoxon"`.
+            If `paired_by` is a tuple, field values are joined with an underscore character ("_").
+        test : {'auto', 'wilcoxon', 'mannwhitneyu', 'kruskal'}, optional
+            Stats test to perform. If `'auto'`, `'mannwhitneyu'` will be chosen if there are two
+            groups of unpaired data. `'wilcoxon'` will be chosen if there are two groups and
+            `paired_by` is specified. `'kruskal'` will be chosen if there are more than 2 groups.
+        metric : {'shannon', 'simpson', 'observed_taxa'}, optional
+            The alpha diversity metric to calculate.
+        rank : {'auto', 'kingdom', 'phylum', 'class', 'order', 'family', 'genus', 'species'}, optional
+            Analysis will be restricted to abundances of taxa at the specified level.
+        alpha : float, optional
+            Threshold to determine statistical significance when `test="kruskal"` (e.g. p < `alpha`).
+            Must be between 0 and 1 (exclusive). If the Kruskal-Wallis p-value is significant, a
+            posthoc Dunn test is performed.
+
+        Returns
+        -------
+        AlphaDiversityStatsResults
+            A dataclass with these attributes:
+            - `test`: the stats test that was performed
+            - `statistic`: the computed test statistic (e.g. U statistic if `test="mannwhitneyu"`)
+            - `pvalue`: the computed p-value
+            - `group_by_variable`: the name of the variable used to group samples by
+            - `groups`: the names of the groups defined by `group_by_variable`
+            - `paired_by_variable`: the name of the variable used to pair samples by (if the data
+              were paired)
+            - `posthoc_df`: `pd.DataFrame` containing Dunn test p-values (if a Dunn test was
+              performed). The index and columns are sorted group names.
+
+        See Also
+        --------
+        scipy.stats.wilcoxon
+        scipy.stats.mannwhitneyu
+        scipy.stats.kruskal
+        scikit_posthocs.posthoc_dunn
+
+        """
+        if not (0.0 < alpha < 1.0):
+            raise StatsException("`alpha` must be between 0 and 1 (exclusive).")
+
+        if not isinstance(group_by, tuple):
```

Comment:
I think it would be good to check if `group_by` is otherwise a string. I can see myself passing a list instead of a tuple here. It might be good to also check for a list and convert it to a tuple. The same applies to `paired_by`